File: paychecker.py
from flask import Blueprint, render_template, jsonify, request
import json

paychecker_bp = Blueprint('paychecker', __name__)

@paychecker_bp.route('/paychecker')
def paychecker():
    """Render the pay checker page"""
    return render_template('paychecker.html')

@paychecker_bp.route('/api/calculate-pay', methods=['POST'])
def calculate_pay():
    """API endpoint to calculate pay based on timesheet data"""
    try:
        data = request.get_json()
        
        # Load rates and allowances from JSON files
        with open('data/minimum_rates.json', 'r') as f:
            rates_data = json.load(f)
        
        # Extract form data
        classification = data.get('classification')
        employment_type = data.get('employmentType')
        timesheet = data.get('timesheet', [])
        conditions = data.get('conditions', {})
        
        # Get base rate for classification
        base_rate = get_base_rate(classification, rates_data)
        
        # Initialize totals
        total_pay = 0
        daily_breakdown = []
        
        # Calculate for each day
        for day_data in timesheet:
            day = day_data.get('day')
            regular_hours = float(day_data.get('regularHours', 0))
            overtime_hours = float(day_data.get('overtimeHours', 0))
            shift_type = day_data.get('shiftType')
            day_conditions = day_data.get('conditions')
            
            # Calculate daily pay
            day_pay = calculate_daily_pay(
                base_rate=base_rate,
                regular_hours=regular_hours,
                overtime_hours=overtime_hours,
                shift_type=shift_type,
                conditions=day_conditions,
                employment_type=employment_type,
                day_of_week=day
            )
            
            daily_breakdown.append({
                'day': day,
                'total': day_pay,
                'details': {
                    'regular': regular_hours * base_rate,
                    'overtime': overtime_hours * base_rate * 1.5,
                    'shift_loading': calculate_shift_loading(shift_type, base_rate, regular_hours),
                    'conditions': calculate_conditions_allowance(day_conditions)
                }
            })
            
            total_pay += day_pay
        
        # Add other allowances
        if conditions.get('inclementWeather'):
            total_pay += 2.85  # Example rate, adjust as needed
        if conditions.get('travelAllowance'):
            total_pay += 17.43  # Example rate, adjust as needed
        
        return jsonify({
            'success': True,
            'total_pay': round(total_pay, 2),
            'daily_breakdown': daily_breakdown
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 400

def get_base_rate(classification, rates_data):
    """Get base hourly rate for classification"""
    # Extract classification level (e.g., "CW/ECW 1" -> "1")
    level = classification.split()[1]
    
    # Find matching rate in rates_data
    classification_key = f"CW/ECW {level}"
    if level == "1":
        classification_key += " - Level a"
    
    if classification_key in rates_data['minimum_rates']:
        weekly_rate = rates_data['minimum_rates'][classification_key]['weekly_rate']
        return weekly_rate / 38  # Convert weekly to hourly
    
    raise ValueError(f"Rate not found for classification: {classification}")

def calculate_daily_pay(base_rate, regular_hours, overtime_hours, shift_type, 
                       conditions, employment_type, day_of_week):
    """Calculate total pay for one day"""
    total = 0
    
    # Regular hours
    if employment_type == 'casual':
        total += regular_hours * base_rate * 1.25  # 25% casual loading
    else:
        total += regular_hours * base_rate
    
    # Overtime
    total += overtime_hours * base_rate * 1.5  # Time and a half
    
    # Shift loadings
    if shift_type == 'nightShift':
        total += regular_hours * base_rate * 0.25  # 25% loading
    elif shift_type == 'afternoonShift':
        total += regular_hours * base_rate * 0.15  # 15% loading
    
    # Weekend rates
    if day_of_week == 'Saturday':
        total *= 1.5  # Time and a half
    elif day_of_week == 'Sunday':
        total *= 2.0  # Double time
    
    return total

def calculate_shift_loading(shift_type, base_rate, hours):
    """Calculate shift loading amount"""
    if shift_type == 'nightShift':
        return hours * base_rate * 0.25
    elif shift_type == 'afternoonShift':
        return hours * base_rate * 0.15
    return 0

def calculate_conditions_allowance(conditions):
    """Calculate conditions allowance amount"""
    allowances = {
        'heightWork': 2.93,
        'confinedSpace': 0.76,
        'tools': 32.47
    }
    return allowances.get(conditions, 0) 

File: static/js/paychecker.js
// Dummy data for testing
const dummyRates = {
    "minimum_rates": {
        "CW/ECW 1": {
            "hourly_rate": 25.88,
            "weekly_rate": 983.44
        },
        "CW/ECW 2": {
            "hourly_rate": 26.76,
            "weekly_rate": 1016.88
        },
        "CW/ECW 3": {
            "hourly_rate": 27.89,
            "weekly_rate": 1059.82
        },
        "CW/ECW 4": {
            "hourly_rate": 29.04,
            "weekly_rate": 1103.52
        },
        "CW/ECW 5": {
            "hourly_rate": 29.85,
            "weekly_rate": 1134.30
        }
    },
    "allowances": {
        "heightWork": 2.93,
        "confinedSpace": 0.76,
        "tools": 32.47,
        "inclementWeather": 2.85,
        "travel": 17.43
    },
    "loadings": {
        "casual": 1.25,
        "nightShift": 1.25,
        "afternoonShift": 1.15,
        "saturday": 1.5,
        "sunday": 2.0,
        "overtime": 1.5
    }
};

// Add this to the template file
document.addEventListener('DOMContentLoaded', function() {
    const form = document.getElementById('payRateForm');
    
    // Pre-fill some test data
    document.querySelector('#classification').value = 'CW/ECW 3';
    document.querySelector('#employmentType').value = 'casual';
    
    // Pre-fill some hours and conditions
    const timesheet = document.querySelectorAll('#timesheet tr');
    
    // Monday - regular day
    timesheet[0].querySelector('.regular-hours').value = 8;
    
    // Tuesday - overtime
    timesheet[1].querySelector('.regular-hours').value = 8;
    timesheet[1].querySelector('.overtime-hours').value = 2;
    
    // Wednesday - night shift
    timesheet[2].querySelector('.regular-hours').value = 8;
    timesheet[2].querySelector('.shift-type').value = 'nightShift';
    
    // Thursday - height work
    timesheet[3].querySelector('.regular-hours').value = 8;
    timesheet[3].querySelector('.conditions').value = 'heightWork';
    
    // Friday - regular day
    timesheet[4].querySelector('.regular-hours').value = 8;
    
    // Saturday - weekend rates
    timesheet[5].querySelector('.regular-hours').value = 4;
    
    // Check some allowances
    document.querySelector('#travelAllowance').checked = true;
    
    // Add form submit handler instead of calculating immediately
    form.addEventListener('submit', function(e) {
        e.preventDefault();  // Prevent form submission
        calculatePay();      // Only calculate when form is submitted
    });
});

function calculatePay() {
    const classification = document.querySelector('#classification').value;
    const baseRate = dummyRates.minimum_rates[classification].hourly_rate;
    const employmentType = document.querySelector('#employmentType').value;
    
    let totalPay = 0;
    const timesheet = document.querySelectorAll('#timesheet tr');
    
    timesheet.forEach((row, index) => {
        const regularHours = parseFloat(row.querySelector('.regular-hours').value) || 0;
        const overtimeHours = parseFloat(row.querySelector('.overtime-hours').value) || 0;
        const shiftType = row.querySelector('.shift-type').value;
        const conditions = row.querySelector('.conditions').value;
        
        // Calculate base pay
        let dayPay = regularHours * baseRate;
        
        // Add casual loading if applicable
        if (employmentType === 'casual') {
            dayPay *= dummyRates.loadings.casual;
        }
        
        // Add overtime
        dayPay += overtimeHours * baseRate * dummyRates.loadings.overtime;
        
        // Add shift loadings
        if (shiftType === 'nightShift') {
            dayPay *= dummyRates.loadings.nightShift;
        } else if (shiftType === 'afternoonShift') {
            dayPay *= dummyRates.loadings.afternoonShift;
        }
        
        // Add weekend rates
        if (index === 5) { // Saturday
            dayPay *= dummyRates.loadings.saturday;
        } else if (index === 6) { // Sunday
            dayPay *= dummyRates.loadings.sunday;
        }
        
        // Add condition allowances
        if (conditions !== 'none') {
            dayPay += dummyRates.allowances[conditions];
        }
        
        totalPay += dayPay;
    });
    
    // Add other allowances
    if (document.querySelector('#inclementWeather').checked) {
        totalPay += dummyRates.allowances.inclementWeather;
    }
    if (document.querySelector('#travelAllowance').checked) {
        totalPay += dummyRates.allowances.travel;
    }
    
    // Update the output display
    document.querySelector('#output').innerHTML = `
        <div class="total-section">
            <span class="total-label">Total Weekly Pay</span>
            <span class="total-amount">$${totalPay.toFixed(2)}</span>
        </div>
    `;
} 